File: src/test/java/ru/org/linux/topic/GroupPermissionServiceTest.java
/*
 * Copyright 1998-2012 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic;

import org.junit.Test;
import ru.org.linux.group.GroupPermissionService;
import ru.org.linux.section.Section;
import ru.org.linux.section.SectionService;
import ru.org.linux.user.User;

import java.sql.ResultSet;
import java.sql.Timestamp;
import java.util.Calendar;
import java.util.Date;

import static org.junit.Assert.*;
import static org.mockito.Mockito.*;

/**
 * Created by IntelliJ IDEA.
 * User: hizel
 * Date: 8/30/11
 * Time: 4:27 PM
 */
public class GroupPermissionServiceTest {
  private final GroupPermissionService permissionService = new GroupPermissionService();

  /**
   * Проверка что пользователь МОЖЕТ удалить топик автором которого он является
   * и прошло меньше часа с момента почтинга
   * @throws Exception
   */
  @Test
  public void isDeletableByUserTest1() throws Exception {
    Calendar calendar = Calendar.getInstance();

    ResultSet resultSet = mock(ResultSet.class);
    when(resultSet.getInt("postscore")).thenReturn(-9999);
    when(resultSet.wasNull()).thenReturn(false);
    calendar.setTime(new Date());
    calendar.add(Calendar.MINUTE, -10);
    // commitdate, lastmod
    when(resultSet.getTimestamp(anyString())).thenReturn(new Timestamp(calendar.getTimeInMillis()));
    // commitby, sectionid, stat1, ua_id,
    when(resultSet.getInt(anyString())).thenReturn(13);
    // gtitle, urlname, message, postip,
    when(resultSet.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSet.getBoolean(anyString())).thenReturn(false);

    when(resultSet.getInt("section")).thenReturn(3); // Галлерея
    when(resultSet.getBoolean("moderate")).thenReturn(true);

    User user = mock(User.class);
    when(user.isModerator()).thenReturn(false);
    when(user.getId()).thenReturn(13);

    Topic message = new Topic(resultSet);

    assertFalse(user.isModerator());
    assertEquals(user.getId(), resultSet.getInt("userid"));
    assertEquals(user.getId(), message.getUid());

    assertTrue(permissionService.isDeletable(message, user));
  }
  /**
   * Проверка что пользователь НЕМОЖЕТ удалить топик автором которого он является
   * и прошло больше часа с момента почтинга
   * @throws Exception
   */
  @Test
  public void isDeletableByUserTest2() throws Exception {
    Calendar calendar = Calendar.getInstance();

    ResultSet resultSet = mock(ResultSet.class);
    when(resultSet.getInt("postscore")).thenReturn(-9999);
    when(resultSet.wasNull()).thenReturn(false);
    calendar.setTime(new Date());
    calendar.add(Calendar.MINUTE, -70);
    // commitdate, lastmod
    when(resultSet.getTimestamp(anyString())).thenReturn(new Timestamp(calendar.getTimeInMillis()));
    // commitby, sectionid, stat1, ua_id,
    when(resultSet.getInt(anyString())).thenReturn(13);
    // gtitle, urlname, message, postip,
    when(resultSet.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSet.getBoolean(anyString())).thenReturn(false);

    when(resultSet.getInt("section")).thenReturn(3); // Галлерея
    when(resultSet.getBoolean("moderate")).thenReturn(true);

    User user = mock(User.class);
    when(user.isModerator()).thenReturn(false);
    when(user.getId()).thenReturn(13);

    Topic message = new Topic(resultSet);

    assertFalse(user.isModerator());
    assertEquals(user.getId(), resultSet.getInt("userid"));
    assertEquals(user.getId(), message.getUid());

    assertFalse(permissionService.isDeletable(message, user));
  }

  /**
   * Проверка что пользователь НЕМОЖЕТ удалить топик автором которого он неявляется
   * и прошло больше часа с момента постинга
   * @throws Exception
   */
  @Test
  public void isDeletableByUserTest3() throws Exception {
    Calendar calendar = Calendar.getInstance();

    ResultSet resultSet = mock(ResultSet.class);
    when(resultSet.getInt("postscore")).thenReturn(-9999);
    when(resultSet.wasNull()).thenReturn(false);
    calendar.setTime(new Date());
    calendar.add(Calendar.MINUTE, -70);
    // commitdate, lastmod
    when(resultSet.getTimestamp(anyString())).thenReturn(new Timestamp(calendar.getTimeInMillis()));
    // commitby, sectionid, stat1, ua_id,
    when(resultSet.getInt(anyString())).thenReturn(13);
    // gtitle, urlname, message, postip,
    when(resultSet.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSet.getBoolean(anyString())).thenReturn(false);

    when(resultSet.getInt("section")).thenReturn(3); // Галлерея
    when(resultSet.getBoolean("moderate")).thenReturn(true);

    User user = mock(User.class);
    when(user.isModerator()).thenReturn(false);
    when(user.getId()).thenReturn(14);

    Topic message = new Topic(resultSet);

    assertFalse(user.isModerator());
    assertFalse(user.getId() == resultSet.getInt("userid"));
    assertFalse(user.getId() == message.getUid());

    assertFalse(permissionService.isDeletable(message, user));
  }

  /**
   * Проверка что пользователь НЕМОЖЕТ удалить топик автором которого он неявляется
   * и прошло больше часа с момента постинга
   * @throws Exception
   */
  @Test
  public void isDeletableByUserTest4() throws Exception {
    Calendar calendar = Calendar.getInstance();

    ResultSet resultSet = mock(ResultSet.class);
    when(resultSet.getInt("postscore")).thenReturn(-9999);
    when(resultSet.wasNull()).thenReturn(false);
    calendar.setTime(new Date());
    calendar.add(Calendar.MINUTE, -5);
    // commitdate, lastmod
    when(resultSet.getTimestamp(anyString())).thenReturn(new Timestamp(calendar.getTimeInMillis()));
    // commitby, sectionid, stat1, ua_id,
    when(resultSet.getInt(anyString())).thenReturn(13);
    // gtitle, urlname, message, postip,
    when(resultSet.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSet.getBoolean(anyString())).thenReturn(false);

    when(resultSet.getInt("section")).thenReturn(3); // Галлерея
    when(resultSet.getBoolean("moderate")).thenReturn(true);

    User user = mock(User.class);
    when(user.isModerator()).thenReturn(false);
    when(user.getId()).thenReturn(14);

    Topic message = new Topic(resultSet);

    assertFalse(user.isModerator());
    assertFalse(user.getId() == resultSet.getInt("userid"));
    assertFalse(user.getId() == message.getUid());

    assertFalse(permissionService.isDeletable(message, user));
  }

  /**
   * Проверка для модератора
   * @throws Exception
   */
  @Test
  public void isDeletableByModeratorTest() throws Exception {
    Calendar calendar = Calendar.getInstance();

    calendar.setTime(new Date());
    calendar.add(Calendar.MONTH, -2);
    long oldTime = calendar.getTimeInMillis();

    calendar.setTime(new Date());
    calendar.add(Calendar.DAY_OF_MONTH, -2);
    long newTime = calendar.getTimeInMillis();


    ResultSet resultSetModerateOld = mock(ResultSet.class);
    when(resultSetModerateOld.getInt("postscore")).thenReturn(-9999);
    when(resultSetModerateOld.wasNull()).thenReturn(false);
    // commitdate, lastmod
    when(resultSetModerateOld.getTimestamp(anyString())).thenReturn(new Timestamp(oldTime));
    // commitby, sectionid, stat1, ua_id,
    when(resultSetModerateOld.getInt(anyString())).thenReturn(13);
    when(resultSetModerateOld.getInt("section")).thenReturn(1);    
    // gtitle, urlname, message, postip,
    when(resultSetModerateOld.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSetModerateOld.getBoolean(anyString())).thenReturn(false);
    when(resultSetModerateOld.getBoolean("moderate")).thenReturn(true);
    when(resultSetModerateOld.getTimestamp("postdate")).thenReturn(new Timestamp(oldTime));


    ResultSet resultSetNotModerateOld = mock(ResultSet.class);
    when(resultSetNotModerateOld.getInt("postscore")).thenReturn(-9999);
    when(resultSetNotModerateOld.wasNull()).thenReturn(false);
    // commitdate, lastmod
    when(resultSetNotModerateOld.getTimestamp(anyString())).thenReturn(new Timestamp(oldTime));
    // commitby, sectionid, stat1, ua_id,
    when(resultSetNotModerateOld.getInt(anyString())).thenReturn(2);
    // gtitle, urlname, message, postip,
    when(resultSetNotModerateOld.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSetNotModerateOld.getBoolean(anyString())).thenReturn(false);
    when(resultSetNotModerateOld.getBoolean("moderate")).thenReturn(false);
    when(resultSetNotModerateOld.getTimestamp("postdate")).thenReturn(new Timestamp(oldTime));


    ResultSet resultSetModerateNew = mock(ResultSet.class);
    when(resultSetModerateNew.getInt("postscore")).thenReturn(-9999);
    when(resultSetModerateNew.wasNull()).thenReturn(false);
    // commitdate, lastmod
    when(resultSetModerateNew.getTimestamp(anyString())).thenReturn(new Timestamp(newTime));
    // commitby, sectionid, stat1, ua_id,
    when(resultSetModerateNew.getInt(anyString())).thenReturn(1);
    // gtitle, urlname, message, postip,
    when(resultSetModerateNew.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSetModerateNew.getBoolean(anyString())).thenReturn(false);
    when(resultSetModerateNew.getBoolean("moderate")).thenReturn(true);
    when(resultSetModerateNew.getTimestamp("postdate")).thenReturn(new Timestamp(newTime));


    ResultSet resultSetNotModerateNew = mock(ResultSet.class);
    when(resultSetNotModerateNew.getInt("postscore")).thenReturn(-9999);
    when(resultSetNotModerateNew.wasNull()).thenReturn(false);
    // commitdate, lastmod
    when(resultSetNotModerateNew.getTimestamp(anyString())).thenReturn(new Timestamp(newTime));
    // commitby, sectionid, stat1, ua_id,
    when(resultSetNotModerateNew.getInt(anyString())).thenReturn(1);
    // gtitle, urlname, message, postip,
    when(resultSetNotModerateNew.getString(anyString())).thenReturn("any");
    // vote, sticky, expired, bbcode, resolved, minor
    when(resultSetNotModerateNew.getBoolean(anyString())).thenReturn(false);
    when(resultSetNotModerateNew.getBoolean("moderate")).thenReturn(false);
    when(resultSetNotModerateNew.getTimestamp("postdate")).thenReturn(new Timestamp(newTime));


    User user = mock(User.class);
    when(user.isModerator()).thenReturn(true);
    when(user.getId()).thenReturn(13);

    // проверка что данные в mock user верные
    assertTrue(user.isModerator());

    Section sectionModerate = mock(Section.class);
    when(sectionModerate.isPremoderated()).thenReturn(true);
    Section sectionNotModerate = mock(Section.class);
    when(sectionNotModerate.isPremoderated()).thenReturn(false);

    SectionService sectionService = mock(SectionService.class);
    when(sectionService.getSection(1)).thenReturn(sectionModerate);
    when(sectionService.getSection(2)).thenReturn(sectionNotModerate);

    permissionService.setSectionService(sectionService);

    // проверка что данные в mock resultSet верные
    assertTrue(resultSetModerateNew.getBoolean("moderate"));
    assertTrue(resultSetModerateOld.getBoolean("moderate"));
    assertFalse(resultSetNotModerateNew.getBoolean("moderate"));
    assertFalse(resultSetNotModerateOld.getBoolean("moderate"));

    assertEquals(0, (new Timestamp(newTime)).compareTo(resultSetModerateNew.getTimestamp("postdate")));
    assertEquals(0, (new Timestamp(oldTime)).compareTo(resultSetModerateOld.getTimestamp("postdate")));
    assertEquals(0, (new Timestamp(newTime)).compareTo(resultSetNotModerateNew.getTimestamp("postdate")));
    assertEquals(0, (new Timestamp(oldTime)).compareTo(resultSetNotModerateOld.getTimestamp("postdate")));


    Topic messageModerateOld = new Topic(resultSetModerateOld);
    Topic messageNotModerateOld = new Topic(resultSetNotModerateOld);
    Topic messageModerateNew = new Topic(resultSetModerateNew);
    Topic messageNotModerateNew = new Topic(resultSetNotModerateNew);

    // проверка что данные в mock message верные
    assertTrue(messageModerateNew.isCommited());
    assertTrue(messageModerateOld.isCommited());
    assertFalse(messageNotModerateNew.isCommited());
    assertFalse(messageNotModerateOld.isCommited());

    assertEquals(0, (new Timestamp(newTime)).compareTo(messageModerateNew.getPostdate()));
    assertEquals(0, (new Timestamp(oldTime)).compareTo(messageModerateOld.getPostdate()));
    assertEquals(0, (new Timestamp(newTime)).compareTo(messageNotModerateNew.getPostdate()));
    assertEquals(0, (new Timestamp(oldTime)).compareTo(messageNotModerateOld.getPostdate()));
    
    // нельзя удалять старые подтвержденные топики в премодерируемом разделе
    assertFalse(permissionService.isDeletable(messageModerateOld, user));
    // можно удалять старые подтвержденные топики в непремодерируемом разделе
//    Assert.assertTrue(permissionService.isDeletableByModerator(messageModerateOld, user));
    // можно удалять старые не подтвержденные топики в премодерируемом разделе
    assertTrue(permissionService.isDeletable(messageNotModerateOld, user));
    // можно удалять старые не подтвержденные топики в непремодерируемом разделе
//    Assert.assertTrue(permissionService.isDeletableByModerator(messageNotModerateOld, user));

    // можно удалять новые подтвержденные топики в премодерируемом разделе
    assertTrue(permissionService.isDeletable(messageModerateNew, user));
    // можно удалять новые подтвержденные топики в непремодерируемом разделе
//    Assert.assertTrue(permissionService.isDeletableByModerator(messageModerateNew, user));
    // можно удалять новые не подтвержденные топики в премодерируемом разделе
    assertTrue(permissionService.isDeletable(messageNotModerateNew, user));
    // можно удалять новые не подтвержденные топики в непремодерируемом разделе
//    Assert.assertTrue(permissionService.isDeletableByModerator(messageNotModerateNew, user));
  }
}

File: src/test/scala/ru/org/linux/topic/GroupPermissionServiceTest.scala
/*
 * Copyright 1998-2026 Linux.org.ru
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 */

package ru.org.linux.topic

import munit.FunSuite
import org.mockito.ArgumentMatchers.anyString
import org.mockito.Mockito.{mock, when}
import ru.org.linux.auth.{AuthorizedSession, IpBlockInfo}
import ru.org.linux.group.GroupPermissionService
import ru.org.linux.section.{Section, SectionScrollModeEnum, SectionService}
import ru.org.linux.user.{Profile, User}

import java.sql.{ResultSet, Timestamp}
import java.util.{Calendar, Date}

class GroupPermissionServiceTest extends FunSuite:
  private def sessionOf(user: User): AuthorizedSession =
    AuthorizedSession(user, user.corrector, user.isModerator, user.isAdministrator,
      Profile.DEFAULT, IpBlockInfo("127.0.0.1"))

  private def createUser(id: Int, moderator: Boolean, corrector: Boolean, administrator: Boolean): User =
    User(nick = "testuser", id = id, canmod = moderator, candel = administrator, anonymous = false,
      corrector = corrector, blocked = false, password = "", score = 0, maxScore = 0,
      photo = null, email = null, fullName = null, unreadEvents = 0, frozenUntil = null,
      activated = true)

  private def mockResultSet(minutesAgo: Int, defaultInt: Int, defaultBool: Boolean,
                             section: Int, moderate: Boolean): ResultSet =
    val calendar = Calendar.getInstance()
    calendar.setTime(new Date())
    calendar.add(Calendar.MINUTE, -minutesAgo)
    val ts = new Timestamp(calendar.getTimeInMillis)

    val resultSet = mock(classOf[ResultSet])
    when(resultSet.getInt("postscore")).thenReturn(-9999)
    when(resultSet.wasNull()).thenReturn(false)
    when(resultSet.getTimestamp(anyString())).thenReturn(ts)
    when(resultSet.getInt(anyString())).thenReturn(defaultInt)
    when(resultSet.getString(anyString())).thenReturn("any")
    when(resultSet.getBoolean(anyString())).thenReturn(defaultBool)

    when(resultSet.getInt("section")).thenReturn(section)
    when(resultSet.getBoolean("moderate")).thenReturn(moderate)
    resultSet

  private def realSection(id: Int, premoderated: Boolean): Section =
    Section(name = "section-" + id, imagepost = true, premoderated = premoderated, id = id,
      votepoll = false, scrollMode = SectionScrollModeEnum.SECTION, topicsRestriction = 0,
      imageAllowed = true)

  /**
    * Проверка, что пользователь МОЖЕТ удалить топик, автором которого он является,
    * и прошло меньше часа с момента постинга
    */
  test("isDeletableByUserTest1"):
    val resultSet = mockResultSet(minutesAgo = 10, defaultInt = 13, defaultBool = false, section = 3,
      moderate = true)
    when(resultSet.getInt("stat1")).thenReturn(0)

    val user = createUser(13, moderator = false, corrector = false, administrator = false)

    val message = Topic.fromResultSet(resultSet)

    assert(!user.isModerator)
    assertEquals(user.id, resultSet.getInt("userid"))
    assertEquals(user.id, message.authorUserId)

    val sectionService = mock(classOf[SectionService])
    when(sectionService.getSection(3)).thenReturn(realSection(3, premoderated = false))

    val permissionService = new GroupPermissionService(sectionService, null, null, null)

    assert(permissionService.isDeletable(message)(using sessionOf(user)))

  /**
    * Проверка, что пользователь НЕ МОЖЕТ удалить топик, автором которого он является,
    * и прошло больше часа с момента постинга
    */
  test("isDeletableByUserTest2"):
    val resultSet = mockResultSet(minutesAgo = 70, defaultInt = 13, defaultBool = false, section = 3,
      moderate = true)

    val user = createUser(13, moderator = false, corrector = false, administrator = false)

    val message = Topic.fromResultSet(resultSet)

    assert(!user.isModerator)
    assertEquals(user.id, resultSet.getInt("userid"))
    assertEquals(user.id, message.authorUserId)

    val sectionService = mock(classOf[SectionService])
    when(sectionService.getSection(3)).thenReturn(realSection(3, premoderated = false))

    val permissionService = new GroupPermissionService(sectionService, null, null, null)

    assert(!permissionService.isDeletable(message)(using sessionOf(user)))

  /**
    * Проверка, что пользователь НЕ МОЖЕТ удалить топик, автором которого он не является,
    * и прошло больше часа с момента постинга
    */
  test("isDeletableByUserTest3"):
    val resultSet = mockResultSet(minutesAgo = 70, defaultInt = 13, defaultBool = false, section = 3,
      moderate = true)

    val user = createUser(14, moderator = false, corrector = false, administrator = false)

    val message = Topic.fromResultSet(resultSet)

    assert(!user.isModerator)
    assertNotEquals(user.id, resultSet.getInt("userid"))
    assertNotEquals(user.id, message.authorUserId)

    val sectionService = mock(classOf[SectionService])
    when(sectionService.getSection(3)).thenReturn(realSection(3, premoderated = false))

    val permissionService = new GroupPermissionService(sectionService, null, null, null)

    assert(!permissionService.isDeletable(message)(using sessionOf(user)))

  /**
    * Проверка, что пользователь НЕ МОЖЕТ удалить топик, автором которого он не является,
    * и прошло меньше часа с момента постинга
    */
  test("isDeletableByUserTest4"):
    val resultSet = mockResultSet(minutesAgo = 5, defaultInt = 13, defaultBool = false, section = 3,
      moderate = true)

    val user = createUser(14, moderator = false, corrector = false, administrator = false)

    val message = Topic.fromResultSet(resultSet)

    assert(!user.isModerator)
    assertNotEquals(user.id, resultSet.getInt("userid"))
    assertNotEquals(user.id, message.authorUserId)

    val sectionService = mock(classOf[SectionService])
    when(sectionService.getSection(3)).thenReturn(realSection(3, premoderated = false))

    val permissionService = new GroupPermissionService(sectionService, null, null, null)

    assert(!permissionService.isDeletable(message)(using sessionOf(user)))

  /** Проверка для модератора */
  test("isDeletableByModeratorTest"):
    val calendar = Calendar.getInstance()

    calendar.setTime(new Date())
    calendar.add(Calendar.MONTH, -2)
    val oldTime = calendar.getTimeInMillis

    calendar.setTime(new Date())
    calendar.add(Calendar.DAY_OF_MONTH, -2)
    val newTime = calendar.getTimeInMillis

    def mockFor(time: Long, defaultInt: Int, moderate: Boolean): ResultSet =
      val resultSet = mock(classOf[ResultSet])
      when(resultSet.getInt("postscore")).thenReturn(-9999)
      when(resultSet.wasNull()).thenReturn(false)
      when(resultSet.getTimestamp(anyString())).thenReturn(new Timestamp(time))
      when(resultSet.getInt(anyString())).thenReturn(defaultInt)
      when(resultSet.getString(anyString())).thenReturn("any")
      when(resultSet.getBoolean(anyString())).thenReturn(false)
      when(resultSet.getBoolean("moderate")).thenReturn(moderate)
      when(resultSet.getTimestamp("postdate")).thenReturn(new Timestamp(time))
      resultSet

    val resultSetModerateOld = mockFor(oldTime, defaultInt = 13, moderate = true)
    when(resultSetModerateOld.getInt("section")).thenReturn(1)

    val resultSetNotModerateOld = mockFor(oldTime, defaultInt = 2, moderate = false)

    val resultSetModerateNew = mockFor(newTime, defaultInt = 1, moderate = true)

    val resultSetNotModerateNew = mockFor(newTime, defaultInt = 1, moderate = false)

    val user = createUser(13, moderator = true, corrector = false, administrator = false)

    // проверка что данные в mock user верные
    assert(user.isModerator)

    val sectionService = mock(classOf[SectionService])
    when(sectionService.getSection(1)).thenReturn(realSection(1, premoderated = true))
    when(sectionService.getSection(2)).thenReturn(realSection(2, premoderated = false))

    val permissionService = new GroupPermissionService(sectionService, null, null, null)

    // проверка, что данные в mock resultSet верные
    assert(resultSetModerateNew.getBoolean("moderate"))
    assert(resultSetModerateOld.getBoolean("moderate"))
    assert(!resultSetNotModerateNew.getBoolean("moderate"))
    assert(!resultSetNotModerateOld.getBoolean("moderate"))

    assertEquals(0, new Timestamp(newTime).compareTo(resultSetModerateNew.getTimestamp("postdate")))
    assertEquals(0, new Timestamp(oldTime).compareTo(resultSetModerateOld.getTimestamp("postdate")))
    assertEquals(0, new Timestamp(newTime).compareTo(resultSetNotModerateNew.getTimestamp("postdate")))
    assertEquals(0, new Timestamp(oldTime).compareTo(resultSetNotModerateOld.getTimestamp("postdate")))

    val messageModerateOld = Topic.fromResultSet(resultSetModerateOld)
    val messageNotModerateOld = Topic.fromResultSet(resultSetNotModerateOld)
    val messageModerateNew = Topic.fromResultSet(resultSetModerateNew)
    val messageNotModerateNew = Topic.fromResultSet(resultSetNotModerateNew)

    // проверка, что данные в mock message верные
    assert(messageModerateNew.commited)
    assert(messageModerateOld.commited)
    assert(!messageNotModerateNew.commited)
    assert(!messageNotModerateOld.commited)

    assertEquals(0, new Timestamp(newTime).compareTo(messageModerateNew.postdate))
    assertEquals(0, new Timestamp(oldTime).compareTo(messageModerateOld.postdate))
    assertEquals(0, new Timestamp(newTime).compareTo(messageNotModerateNew.postdate))
    assertEquals(0, new Timestamp(oldTime).compareTo(messageNotModerateOld.postdate))

    // нельзя удалять старые подтверждённые топики в премодерируемом разделе
    assert(!permissionService.isDeletable(messageModerateOld)(using sessionOf(user)))
    // можно удалять старые не подтверждённые топики в премодерируемом разделе
    assert(permissionService.isDeletable(messageNotModerateOld)(using sessionOf(user)))

    // можно удалять новые подтверждённые топики в премодерируемом разделе
    assert(permissionService.isDeletable(messageModerateNew)(using sessionOf(user)))
    // можно удалять новые не подтверждённые топики в премодерируемом разделе
    assert(permissionService.isDeletable(messageNotModerateNew)(using sessionOf(user)))